http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java

```diff
@@ -1407,49 +1408,47 @@ protected void doOnError(Throwable t) {
                         final Supplier<CompletableFuture<?>> supplier = ContinuationArgumentBinder.extractContinuationCompletableFutureSupplier(incomingRequest);
-                        Object suspendedBody;
                         if (isKotlinCoroutineSuspended(body)) {
-                            if (isKotlinFunctionReturnTypeUnit) {
-                                suspendedBody = Completable.create(emitter -> {
-                                    CompletableFuture<?> f = supplier.get();
-                                    f.whenComplete((BiConsumer<Object, Throwable>) (o, throwable) -> {
-                                        if (throwable != null) {
-                                            emitter.onError(throwable);
-                                        } else {
-                                            emitter.onComplete();
-                                        }
-                                    });
-                                });
-                            } else {
-                                suspendedBody = Single.create(emitter -> {
-                                    CompletableFuture<?> f = supplier.get();
-                                    f.whenComplete((BiConsumer<Object, Throwable>) (o, throwable) -> {
-                                        if (throwable != null) {
-                                            emitter.onError(throwable);
+                            return Single.<MutableHttpResponse<?>>create(emitter -> {
+                                CompletableFuture<?> f = supplier.get();
+                                f.whenComplete((o, throwable) -> {
+                                    if (throwable != null) {
+                                        emitter.onError(throwable);
+                                    } else {
+                                        MutableHttpResponse<?> response;
+                                        if (o instanceof MutableHttpResponse) {
+                                            response = (MutableHttpResponse<?>) o;
                                         } else {
-                                            emitter.onSuccess(o);
+                                            response = forStatus(routeMatch.getAnnotationMetadata(), defaultHttpStatus);
+                                            if (!isKotlinFunctionReturnTypeUnit) {
+                                                response = response.body(o);
+                                            }
                                         }
-                                    });
+                                        response.setAttribute(HttpAttributes.ROUTE_MATCH, finalRoute);
+                                        emitter.onSuccess(response);
+                                    }
                                 });
-                            }
+                            }).toFlowable();
```

Comment:
Isn't it just more efficient to use `Flowable.create` instead of `Single.create` if we are just converting it to a flowable afterwards? Only different in the code would be you would have to call `onNext(..)` then `onComplete()`

Reply:
Probably - I'll update it